maven-resources-plugin NoSuchFileException when copying resources with targetPath

## Description

Building `tomcat-taglibs-rdc` with Maven from the `maven-4.0.x` branch causes `maven-resources-plugin:3.3.1` to fail with a `NoSuchFileException` when copying resource files that use `<targetPath>`.

The same build succeeds with Maven 3.x.

## Error

```
[ERROR] Failed to execute goal
  org.apache.maven.plugins:maven-resources-plugin:3.3.1:resources (default-resources)
  on project taglibs-rdc:
  copying .../src/main/resources/org/apache/taglibs/rdc/resources/RDCBundle_en_US.properties
  to /org/apache/taglibs/rdc/resources/RDCBundle_en_US.properties
  failed with NoSuchFileException -> [Help 1]
```

Note the target path starts with `/org/apache/...` — this appears to be a `<targetPath>` configuration in the resource definition. Maven 4 may be resolving the target path differently, causing the directory to not exist when the copy is attempted.

## Reproducer

```bash
git clone https://github.com/apache/tomcat-taglibs-rdc.git
cd tomcat-taglibs-rdc
mvn -B -e clean package -DskipTests
```

## Environment

- Maven: built from `maven-4.0.x` branch (bootstrapped with 4.0.0-rc-5)
- JDK: 17 (Temurin)

## Context

Found during [Maven 4 compatibility testing](https://github.com/gnodet/maven4-testing/issues/20677).

_Claude Code on behalf of Guillaume Nodet_

--- a/impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultSourceRoot.java
+++ b/impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultSourceRoot.java
@@ -108,7 +108,12 @@ public record DefaultSourceRoot(
         this.includes = (includes != null) ? List.copyOf(includes) : List.of();
         this.excludes = (excludes != null) ? List.copyOf(excludes) : List.of();
         this.stringFiltering = stringFiltering;
-        this.targetPathOrNull = (targetPathOrNull != null) ? targetPathOrNull.normalize() : null;
+        if (targetPathOrNull != null) {
+            Path normalized = targetPathOrNull.normalize();
+            this.targetPathOrNull = normalized.toString().isEmpty() ? null : normalized;
+        } else {
+            this.targetPathOrNull = null;
+        }
         this.enabled = enabled;
     }
 

--- a/impl/maven-impl/src/test/java/org/apache/maven/impl/DefaultSourceRootTest.java
+++ b/impl/maven-impl/src/test/java/org/apache/maven/impl/DefaultSourceRootTest.java
@@ -229,6 +229,34 @@ public class DefaultSourceRootTest {
         assertFalse(targetPath.isPresent(), "targetPath should be empty for empty string");
     }
 
+    /** GH-12306: {@code targetPath="."} normalizes to an empty path and must be treated as absent. */
+    @Test
+    void testHandlesDotTargetPathFromResource() {
+        Resource resource = Resource.newBuilder()
+                .directory("src/main/resources")
+                .targetPath(".")
+                .build();
+
+        DefaultSourceRoot sourceRoot = new DefaultSourceRoot(Path.of("myproject"), ProjectScope.MAIN, resource);
+
+        assertFalse(sourceRoot.targetPath().isPresent(), "targetPath \".\" should be treated as absent");
+    }
+
+    /** GH-12306: {@code targetPath="./subdir"} normalizes to {@code "subdir"} and must be preserved. */
+    @Test
+    void testHandlesDotRelativeTargetPathFromResource() {
+        Resource resource = Resource.newBuilder()
+                .directory("src/main/resources")
+                .targetPath("./subdir")
+                .build();
+
+        DefaultSourceRoot sourceRoot = new DefaultSourceRoot(Path.of("myproject"), ProjectScope.MAIN, resource);
+
+        assertTrue(
+                sourceRoot.targetPath().isPresent(), "targetPath \"./subdir\" should be present after normalization");
+        assertEquals(Path.of("subdir"), sourceRoot.targetPath().orElseThrow());
+    }
+
     /*MNG-11062*/
     @Test
     void testHandlesPropertyPlaceholderInTargetPath() {

